File: step_01_basics/1.1_logic_building/P019_pattern_19_symmetric_void_pattern/solution.py
"""
Problem: Pattern 19 - Symmetric Void Pattern
Difficulty: EASY | XP: 10

Print symmetric void (hourglass of spaces between stars). For n=5:
**********
****  ****
***    ***
**      **
*        *
*        *
**      **
***    ***
****  ****
**********
"""
from typing import List, Optional


# ============================================================
# APPROACH 1: BRUTE FORCE -- Separate Top and Bottom Halves
# Time: O(N^2)  |  Space: O(1)
# Print upper half (n rows) then lower half (n rows) with
# three inner loops each: left stars, spaces, right stars.
# ============================================================
def brute_force(n: int) -> None:
    # Upper half (rows 0..n-1)
    for i in range(n):
        # left stars: n - i
        for j in range(n - i):
            print("*", end="")
        # middle spaces: 2 * i
        for j in range(2 * i):
            print(" ", end="")
        # right stars: n - i
        for j in range(n - i):
            print("*", end="")
        print()
    # Lower half (rows 0..n-1), mirror of upper
    for i in range(n):
        # left stars: i + 1
        for j in range(i + 1):
            print("*", end="")
        # middle spaces: 2 * (n - 1 - i)
        for j in range(2 * (n - 1 - i)):
            print(" ", end="")
        # right stars: i + 1
        for j in range(i + 1):
            print("*", end="")
        print()


# ============================================================
# APPROACH 2: OPTIMAL -- Unified Loop with Conditional
# Time: O(N^2)  |  Space: O(1)
# Single loop of 2*n rows. Compute stars and spaces from row index.
# ============================================================
def optimal(n: int) -> None:
    total_rows = 2 * n
    for i in range(total_rows):
        if i < n:
            stars = n - i
            spaces = 2 * i
        else:
            stars = i - n + 1
            spaces = 2 * (2 * n - 1 - i)
        print("*" * stars + " " * spaces + "*" * stars)


# ============================================================
# APPROACH 3: BEST -- String Multiplication, Single Print
# Time: O(N^2)  |  Space: O(N^2)
# Build all rows as strings, join, single print.
# ============================================================
def best(n: int) -> None:
    lines = []
    for i in range(2 * n):
        if i < n:
            stars = n - i
            spaces = 2 * i
        else:
            stars = i - n + 1
            spaces = 2 * (2 * n - 1 - i)
        lines.append("*" * stars + " " * spaces + "*" * stars)
    print("\n".join(lines))


if __name__ == "__main__":
    N = 5
    print("=== Pattern 19 - Symmetric Void Pattern ===")
    print("--- Brute Force ---")
    brute_force(N)
    print("--- Optimal ---")
    optimal(N)
    print("--- Best ---")
    best(N)

File: step_01_basics/1.1_logic_building/P019_pattern_19_symmetric_void_pattern/Solution.java
/**
 * Problem: Pattern 19 - Symmetric Void Pattern
 * Difficulty: EASY | XP: 10
 *
 * Print symmetric void (hourglass of spaces between stars). For n=5:
 * **********
 * ****  ****
 * ***    ***
 * **      **
 * *        *
 * *        *
 * **      **
 * ***    ***
 * ****  ****
 * **********
 *
 * @author DSA_Nova
 */
public class Solution {

    // ============================================================
    // APPROACH 1: BRUTE FORCE -- Separate Top and Bottom Halves
    // Time: O(N^2)  |  Space: O(1)
    // Three inner loops per row: left stars, spaces, right stars.
    // ============================================================
    static class BruteForce {
        public static void print(int n) {
            // Upper half
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n - i; j++) System.out.print("*");
                for (int j = 0; j < 2 * i; j++) System.out.print(" ");
                for (int j = 0; j < n - i; j++) System.out.print("*");
                System.out.println();
            }
            // Lower half
            for (int i = 0; i < n; i++) {
                for (int j = 0; j <= i; j++) System.out.print("*");
                for (int j = 0; j < 2 * (n - 1 - i); j++) System.out.print(" ");
                for (int j = 0; j <= i; j++) System.out.print("*");
                System.out.println();
            }
        }
    }

    // ============================================================
    // APPROACH 2: OPTIMAL -- Unified Loop with Conditional
    // Time: O(N^2)  |  Space: O(1)
    // Single loop of 2*n rows, compute stars/spaces from index.
    // ============================================================
    static class Optimal {
        public static void print(int n) {
            for (int i = 0; i < 2 * n; i++) {
                int stars, spaces;
                if (i < n) {
                    stars = n - i;
                    spaces = 2 * i;
                } else {
                    stars = i - n + 1;
                    spaces = 2 * (2 * n - 1 - i);
                }
                for (int j = 0; j < stars; j++) System.out.print("*");
                for (int j = 0; j < spaces; j++) System.out.print(" ");
                for (int j = 0; j < stars; j++) System.out.print("*");
                System.out.println();
            }
        }
    }

    // ============================================================
    // APPROACH 3: BEST -- StringBuilder, Single Print
    // Time: O(N^2)  |  Space: O(N^2)
    // Build entire pattern, single I/O call.
    // ============================================================
    static class Best {
        public static void print(int n) {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < 2 * n; i++) {
                int stars, spaces;
                if (i < n) {
                    stars = n - i;
                    spaces = 2 * i;
                } else {
                    stars = i - n + 1;
                    spaces = 2 * (2 * n - 1 - i);
                }
                for (int j = 0; j < stars; j++) sb.append('*');
                for (int j = 0; j < spaces; j++) sb.append(' ');
                for (int j = 0; j < stars; j++) sb.append('*');
                if (i < 2 * n - 1) sb.append('\n');
            }
            System.out.println(sb.toString());
        }
    }

    public static void main(String[] args) {
        int n = 5;
        System.out.println("=== Pattern 19 - Symmetric Void Pattern ===");
        System.out.println("--- Brute Force ---");
        BruteForce.print(n);
        System.out.println("--- Optimal ---");
        Optimal.print(n);
        System.out.println("--- Best ---");
        Best.print(n);
    }
}
